File: frontend/src/lib/table/custom-export.ts
/**
 * Client mirror of core/table/custom_export.py::overridden_table — used by
 * the entry layout dialog and the json-preview call. Display-only; the
 * backend re-applies overrides itself at /exports/run time, so drift here
 * cannot corrupt an export (same stance as table/export-layout.ts).
 */
import type { ColumnOverride, ExportEntry, TableDefinition } from '$lib/api/types';

export function applyEntryOverrides(defn: TableDefinition, entry: ExportEntry): TableDefinition {
	const byIndex = new Map<number, ColumnOverride>();
	for (const ov of entry.columns) {
		if (ov.index >= 0 && ov.index < defn.columns.length && !byIndex.has(ov.index))
			byIndex.set(ov.index, ov);
	}
	return {
		...defn,
		columns: defn.columns.map((col, i) => {
			const ov = byIndex.get(i);
			return { ...col, export: ov?.export ?? null, json_export: ov?.json_export ?? null };
		}),
		export_order: [...entry.export_order],
		show_row_numbers: entry.show_row_numbers,
		export_row_number: entry.export_row_number ?? null,
		json_split: entry.json_split ?? null
	};
}

export function overridesFromDefinition(
	defn: TableDefinition
): Pick<
	ExportEntry,
	'columns' | 'export_order' | 'show_row_numbers' | 'export_row_number' | 'json_split'
> {
	const columns: ColumnOverride[] = [];
	defn.columns.forEach((col, index) => {
		if (col.export != null || col.json_export != null)
			columns.push({
				index,
				export: col.export ?? null,
				json_export: col.json_export ?? null
			});
	});
	return {
		columns,
		export_order: [...defn.export_order],
		show_row_numbers: defn.show_row_numbers,
		export_row_number: defn.export_row_number ?? null,
		json_split: defn.json_split ?? null
	};
}

export function entryForTable(tableId: string, defn: TableDefinition, name: string): ExportEntry {
	return {
		source: { ref: tableId },
		name,
		format: 'xlsx',
		...overridesFromDefinition(defn)
	};
}

File: src/data_rover/core/table/custom_export.py
"""The `kind='custom_export'` artifact payload: a named collection of table
exports whose presentation lives IN the artefact (spec §3.3).

`overridden_table` is the override mechanism: a copy of the table definition
whose PRESENTATION fields are restated from an entry. RENDER ONLY — the copy
feeds `export_layout`/`export_header`/`export_definition`/`render_json` and
nothing else; evaluation keeps the original definition so cell values, row
order and script cache keys are independent of the artefact (the
`export_definition` boundary, applied from a different source).

Spec: docs/superpowers/specs/2026-08-13-table-export-split-and-custom-export-design.md
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, TypeAdapter

from .schema import (
    ColumnExportOptions,
    JsonColumnOptions,
    JsonSplitOptions,
    RowNumberExportOptions,
    TableDefinition,
)


class TableRef(BaseModel):
    """Serialized as a dict under the literal key `"ref"` — the shape
    artifact_kinds.extract_refs's generic walk already understands, so the
    bundle deps closure and id rewriting need zero per-kind code."""

    ref: str


class ColumnOverride(BaseModel):
    """Presentation override for ONE definition column, keyed by index.
    Indices drift-normalize against the current definition at apply time
    (out-of-range dropped, duplicates first-wins) — `normalized_order`'s
    stance: a stale override left by a column remove must not block an
    export."""

    index: int = Field(ge=0)
    export: ColumnExportOptions | None = None
    json_export: JsonColumnOptions | None = None


class ExportEntry(BaseModel):
    source: TableRef
    #: Output base name in the zip; "" falls back to the table's name.
    name: str = ""
    format: Literal["xlsx", "json"] = "xlsx"
    columns: list[ColumnOverride] = Field(default_factory=list)
    export_order: list[int] = Field(default_factory=list)
    show_row_numbers: bool = False
    export_row_number: RowNumberExportOptions | None = None
    json_split: JsonSplitOptions | None = None


class CustomExportDefinition(BaseModel):
    schema_version: int = 1
    entries: list[ExportEntry] = Field(default_factory=list)


CUSTOM_EXPORT_ADAPTER: TypeAdapter[CustomExportDefinition] = TypeAdapter(
    CustomExportDefinition
)


def overridden_table(defn: TableDefinition, entry: ExportEntry) -> TableDefinition:
    """A copy of `defn` whose presentation restates `entry`.

    Columns the entry does not mention get DEFAULT presentation (`export=None`
    -> include follows `hidden`; `json_export=None`), deliberately NOT the
    table's own standalone settings — the two config sets never bleed into
    each other in either direction. Structural fields (sources, modes,
    `hidden`, filters, row source) are untouched: an entry restates how a
    table RENDERS, never what it computes.
    """
    by_index: dict[int, ColumnOverride] = {}
    for ov in entry.columns:
        if 0 <= ov.index < len(defn.columns) and ov.index not in by_index:
            by_index[ov.index] = ov
    columns = []
    for i, col in enumerate(defn.columns):
        override = by_index.get(i)
        columns.append(
            col.model_copy(
                update={
                    "export": override.export if override is not None else None,
                    "json_export": override.json_export
                    if override is not None
                    else None,
                }
            )
        )
    return defn.model_copy(
        update={
            "columns": columns,
            "export_order": list(entry.export_order),
            "show_row_numbers": entry.show_row_numbers,
            "export_row_number": entry.export_row_number,
            "json_split": entry.json_split,
        }
    )
